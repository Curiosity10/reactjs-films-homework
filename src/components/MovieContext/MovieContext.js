import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { getPopularMovies } from '../../utils/tbdbApiService';

export const MovieContext = React.createContext();

export const MovieContextProvider = ({ children }) => {
  const [layout, setLayout] = useState('grid');
  const initialState = {
    movies: getPopularMovies(),
    layout,
    changeLayout: setLayout,
  };
  return (
    <MovieContext.Provider value={initialState}>
      {children}
    </MovieContext.Provider>
  );
};

MovieContextProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};
