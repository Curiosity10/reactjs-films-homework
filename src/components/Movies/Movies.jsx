import React, { useState } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames/bind';
import styles from './Movies.scss';
import Grid from './components/Grid';
import Table from './components/Table';
import Spinner from '../Spinner';
import TrailerModal from './components/TrailerModal';

const cx = cn.bind(styles);

const Movies = ({
  isLoading, movies, layout, src, page,
}) => {
  const [isModalOpen, setModalOpen] = useState(false);
  const isMoviesExist = movies.length !== 0;
  let moviesList;

  if (!isMoviesExist) {
    moviesList = <p className={styles.notFound}>Sorry, no movies found.</p>;
  } else if (layout === 'grid') {
    moviesList = <Grid toggleModal={setModalOpen} movies={movies} />;
  } else if (layout === 'table') {
    moviesList = <Table toggleModal={setModalOpen} movies={movies} />;
  }

  return (
    <>
      {
        <main className={
          cx({
            container: true,
            gridMovieList: (layout === 'grid' && isMoviesExist),
            tableMovieList: (layout === 'table' && isMoviesExist),
          })}
        >
          { moviesList }
          {page > 1 && <a className={styles.toTopBtn} type="button" href="#menu">&#x2B06;</a>}
        </main>
      }
      {isModalOpen && (
        <TrailerModal
          isLoading={isLoading}
          toggleModal={setModalOpen}
          src={src}
        />
      )}
      { isLoading && <Spinner /> }
    </>
  );
};

Movies.propTypes = {
  isLoading: PropTypes.bool,
  movies: PropTypes.arrayOf(PropTypes.object),
  layout: PropTypes.string,
  src: PropTypes.string,
  page: PropTypes.number,
};

Movies.defaultProps = {
  movies: [],
  isLoading: false,
  layout: 'Grid',
  src: '',
  page: 1,
};

export default Movies;
