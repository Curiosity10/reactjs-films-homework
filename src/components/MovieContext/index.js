import { MovieContext, MovieContextProvider } from './MovieContext';

export { MovieContext, MovieContextProvider };
