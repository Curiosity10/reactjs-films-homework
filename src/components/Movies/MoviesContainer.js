import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Movies from './Movies';
import {
  fetchTopRatedMovies,
  fetchLatestMovies,
  fetchUpcomingMovies,
  fetchSearchMovies,
  fetchMoviesByGenre,
} from '../../modules/app/appActions';
import layoutSelector from '../../modules/ui/uiSelectors';
import {
  filterSelector,
  currentPageSelector,
  isLoadingSelector,
  genreIdSelector,
  searchQuerySelector,
  hasMorePagesSelector,
  videoSrcSelector,
  getMoviesSelector,
} from '../../modules/app/appSelectors';
import { useFetchOnScroll } from '../../hooks';

const MovieContainer = ({
  isLoading, filter, currentPage, genreId,
  searchQuery, movies, layout, getTopRatedMovies,
  getLatestMovies, getUpcomingMovies, getSearchMovies,
  getMoviesByGenre, videoSrc, hasMorePages,
}) => {
  useFetchOnScroll(
    isLoading, hasMorePages, filter, currentPage,
    getLatestMovies, getTopRatedMovies, getUpcomingMovies,
    getMoviesByGenre, genreId, getSearchMovies, searchQuery,
  );
  return (
    <Movies
      page={currentPage}
      videoSrc={videoSrc}
      isLoading={isLoading}
      movies={movies}
      layout={layout}
    />
  );
};

const mapStateToProps = state => ({
  movies: getMoviesSelector(state),
  layout: layoutSelector(state),
  currentPage: currentPageSelector(state),
  isLoading: isLoadingSelector(state),
  filter: filterSelector(state),
  genreId: genreIdSelector(state),
  searchQuery: searchQuerySelector(state),
  hasMorePages: hasMorePagesSelector(state),
  videoSrc: videoSrcSelector(state),
});

const actions = {
  getTopRatedMovies: fetchTopRatedMovies,
  getLatestMovies: fetchLatestMovies,
  getUpcomingMovies: fetchUpcomingMovies,
  getSearchMovies: fetchSearchMovies,
  getMoviesByGenre: fetchMoviesByGenre,
};

MovieContainer.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  layout: PropTypes.string.isRequired,
  currentPage: PropTypes.number.isRequired,
  filter: PropTypes.string.isRequired,
  genreId: PropTypes.string,
  searchQuery: PropTypes.string,
  videoSrc: PropTypes.string,
  hasMorePages: PropTypes.bool.isRequired,
  movies: PropTypes.arrayOf(PropTypes.object),
  getTopRatedMovies: PropTypes.func.isRequired,
  getLatestMovies: PropTypes.func.isRequired,
  getUpcomingMovies: PropTypes.func.isRequired,
  getSearchMovies: PropTypes.func.isRequired,
  getMoviesByGenre: PropTypes.func.isRequired,
};

MovieContainer.defaultProps = {
  genreId: null,
  searchQuery: '',
  videoSrc: '',
  movies: [],
};

export default connect(mapStateToProps, actions)(MovieContainer);
