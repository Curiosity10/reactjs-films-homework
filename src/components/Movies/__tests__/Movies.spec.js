import React from 'react';
import renderer from 'react-test-renderer';
import Movies from '..';
import { MovieContextProvider } from '../../MovieContext';

describe('Default Movies component renders correctly', () => {
  it('Grid renders correctly', () => {
    const tree = renderer.create(
      <MovieContextProvider>
        <Movies />
      </MovieContextProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
