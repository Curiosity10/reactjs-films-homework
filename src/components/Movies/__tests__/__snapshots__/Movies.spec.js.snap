// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies component renders correctly Movies grid renders correctly 1`] = `
<React.Fragment>
  <main
    className="container gridMovieList"
  >
    <Grid
      movies={
        Array [
          Object {
            "title": "first movie",
          },
          Object {
            "title": "second movie",
          },
        ]
      }
      toggleModal={[Function]}
    />
    <a
      className="toTopBtn"
      href="#top"
      type="button"
    >
      â¬†
    </a>
  </main>
</React.Fragment>
`;

exports[`Movies component renders correctly Movies not found message renders correctly 1`] = `
<React.Fragment>
  <main
    className="container"
  >
    <p
      className="notFound"
    >
      Sorry, no movies found.
    </p>
  </main>
</React.Fragment>
`;
